File: src/Test/StateMonadFFI.php
<?php
$exports['runStateMonadFFI'] = function($input) {
    $pure = function($value) {
        return function($state) use ($value) { return (object)['value' => $value, 'state' => $state]; };
    };
    $bind = function($action, $next) {
        return function($state) use ($action, $next) {
            $first = $action($state);
            return $next($first->value)($first->state);
        };
    };
    $get = function($state) { return (object)['value' => $state, 'state' => $state]; };
    $put = function($state) {
        return function($_) use ($state) { return (object)['value' => null, 'state' => $state]; };
    };
    $modify = function($f) use ($bind, $get, $put) {
        return $bind($get, function($state) use ($f, $put) { return $put($f($state)); });
    };
    $chain = function($depth) use (&$chain, $pure, $bind, $modify) {
        if ($depth === 0) return $pure(null);
        return $bind($modify(function($value) { return $value + 1; }),
            function($_) use (&$chain, $depth) { return $chain($depth - 1); });
    };
    $total = 0;
    for ($i = 0; $i < 20; $i++) $total += $chain((int)$input)(0)->state;
    return $total;
};
return $exports;

File: src/Test/StateMonadFFI.cc
#include "purescript.h"
#include <functional>
#include <memory>

namespace {
    struct Unit {};
    template <class S, class A> struct StateResult { A val; S state; };
    template <class S, class A> using StateFn = std::function<StateResult<S, A>(S)>;
    template <class S, class A> using State = std::shared_ptr<const StateFn<S, A>>;

    template <class S, class A, class B>
    State<S, B> bindState(State<S, A> action, std::function<State<S, B>(A)> next) {
        return std::make_shared<StateFn<S, B>>([action, next](S state) {
            const auto result = (*action)(state);
            return (*next(result.val))(result.state);
        });
    }

    template <class S, class A>
    State<S, A> pureState(A value) {
        return std::make_shared<StateFn<S, A>>([value](S state) {
            return StateResult<S, A>{value, state};
        });
    }

    template <class S>
    State<S, S> get() {
        return std::make_shared<StateFn<S, S>>([](S state) {
            return StateResult<S, S>{state, state};
        });
    }

    template <class S>
    State<S, Unit> put(S nextState) {
        return std::make_shared<StateFn<S, Unit>>([nextState](S) {
            return StateResult<S, Unit>{Unit{}, nextState};
        });
    }

    template <class S>
    State<S, Unit> modify(std::function<S(S)> f) {
        return bindState<S, S, Unit>(get<S>(), [f](S state) { return put<S>(f(state)); });
    }

    State<int, Unit> chainModifications(int depth) {
        if (depth == 0) return pureState<int, Unit>(Unit{});
        return bindState<int, Unit, Unit>(modify<int>([](int x) { return x + 1; }),
            [depth](Unit) { return chainModifications(depth - 1); });
    }
}

FOREIGN_BEGIN(Test_StateMonadFFI)
exports["runStateMonadFFI"] = [](const boxed& in) -> boxed {
    // This FFI wrapper passes the chain depth (60), not the repetition count.
    const int depth = unbox<int>(in);
    int result = 0;
    for (int repetition = 0; repetition < 20; ++repetition) {
        result += (*chainModifications(depth))(0).state;
    }
    return result;
};
FOREIGN_END
